File: update_data.js
const fs = require('fs');
const sizeOf = require('image-size');
const path = require('path');

let content = fs.readFileSync('data/portfolioData.js', 'utf8');

// Find all categories
const categoryRegex = /folder:\s*"([^"]+)"[\s\S]*?projects:\s*\[([\s\S]*?)\]\n    \}/g;

let updatedContent = content;

let match;
while ((match = categoryRegex.exec(content)) !== null) {
  const fullMatch = match[0];
  const folder = match[1];
  const projectsStr = match[2];
  
  const projectsRegex = /images:\s*\[([\s\S]*?)\]/g;
  let updatedProjectsStr = projectsStr;
  
  let pMatch;
  while ((pMatch = projectsRegex.exec(projectsStr)) !== null) {
    const imagesMatch = pMatch[0];
    const imagesStr = pMatch[1];
    
    // Check if already updated to objects
    if (imagesStr.includes('{')) continue;
    
    const strings = [...imagesStr.matchAll(/"([^"]+)"/g)].map(m => m[1]);
    
    const objects = strings.map(filename => {
      let imgPath = path.join('assets', 'images', folder, filename);
      let width = 800, height = 600;
      try {
        const dimensions = sizeOf(imgPath);
        width = dimensions.width;
        height = dimensions.height;
      } catch (e) {
        console.log("Error reading", imgPath);
      }
      return `\n            { filename: "${filename}", width: ${width}, height: ${height} }`;
    });
    
    const newImagesMatch = `images: [${objects.join(',')}\n          ]`;
    updatedProjectsStr = updatedProjectsStr.replace(imagesMatch, newImagesMatch);
  }
  
  const updatedFullMatch = fullMatch.replace(projectsStr, updatedProjectsStr);
  updatedContent = updatedContent.replace(fullMatch, updatedFullMatch);
}

fs.writeFileSync('data/portfolioData.js', updatedContent);
console.log('Successfully updated portfolioData.js');

File: update_data.py
import os
import re
from PIL import Image

file_path = 'data/portfolioData.js'
with open(file_path, 'r', encoding='utf-8') as f:
    content = f.read()

def repl_projects(match):
    folder = match.group(1)
    projects_str = match.group(2)
    
    def repl_images(img_match):
        img_str = img_match.group(1)
        if '{' in img_str:
            return img_match.group(0)
            
        filenames = re.findall(r'"([^"]+)"', img_str)
        objects = []
        for filename in filenames:
            img_path = os.path.join('assets', 'images', folder, filename)
            width, height = 800, 600
            try:
                with Image.open(img_path) as img:
                    width, height = img.size
            except Exception as e:
                print(f"Error reading {img_path}: {e}")
                
            objects.append(f'\n            {{ filename: "{filename}", width: {width}, height: {height} }}')
            
        return 'images: [' + ','.join(objects) + '\n          ]'
        
    updated_projects = re.sub(r'images:\s*\[([\s\S]*?)\]', repl_images, projects_str)
    return match.group(0).replace(projects_str, updated_projects)

updated_content = re.sub(r'folder:\s*"([^"]+)"[\s\S]*?projects:\s*\[([\s\S]*?)\]\n    \}', repl_projects, content)

with open(file_path, 'w', encoding='utf-8') as f:
    f.write(updated_content)

print("Done")
